### PR TITLE
fix: render dropped file attachments as file pills

### PR DESCRIPTION
**Pull request by maccman**
## Problem

Dragging a file (e.g. a PDF) onto a note saved it into `assets/` and inserted a working `[name](assets/…)` link — but it rendered as an ordinary hyperlink instead of meowdown's rounded file pill. Pill rendering is opt-in: meowdown only renders a link as a pill when the host claims it via `resolveFileLink`, and Reflect never passed that hook (nor `resolveFileInfo` / `onFileClick`).

## Changes

- **`use-asset-persistence.ts`** — exports `resolveAssetFileLink`, a pure resolver that claims safe graph-relative `assets/…` links (same vetting as `resolveAssetOpenPath`), and adds `resolveFileInfo`, which supplies the size shown on a pill: seeded from each save (the size is already in hand) and backfilled by one shared `assets/` `dir_list` per burst of pills, cached per graph session.
- **`note-editor.tsx`** — new `resolveFileLink` / `resolveFileInfo` props passed through to `MeowdownEditor`, plus an `onFileClick` that routes a pill click through the same href handling as a link click (asset opener → deep link → OS opener), so attachments keep opening in the OS default app.
- **`note-pane.tsx`** — wires both into the editor.

## Testing

- `pnpm typecheck` and `pnpm lint` pass (only pre-existing warnings).
- New unit tests: `resolveAssetFileLink` vetting, `resolveFileInfo` (single listing per burst, save-seeded sizes, unsafe/remote/missing/no-session declines), and prop/click plumbing in `note-editor.test.tsx` — 36 tests pass across the two files.
- Verified in the browser preview (`?platform=ios`, in-memory bridge): an `assets/q3-report.pdf` link renders as a rounded pill (`data-file-kind="pdf"`, icon + name, 999px radius) in the daily note.

## Notes

- Read-only `MarkdownView` surfaces (backlink snippets, previews) still render asset links as plain links; claiming pills there is a possible follow-up.
- The browser dev bridge lacks `asset_upload_*`, so drag-drop itself was verified only by unit tests; pill rendering was verified live.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **Low Risk**
> Editor UX and asset path vetting only; reuses existing safe `assets/` checks and link-open behavior, with session-scoped caching covered by tests.
> 
> **Overview**
> Dropped or pasted non-image files already saved to `assets/` and inserted as `[name](assets/…)` links, but they still looked like plain hyperlinks because Reflect never opted into meowdown’s file-pill rendering.
> 
> **`use-asset-persistence`** now exports **`resolveAssetFileLink`** (claims only vetted graph-relative `assets/…` paths, same rules as asset open) and **`resolveFileInfo`**, which shows pill sizes from the in-flight save and otherwise batches one `assets/` directory listing per graph session, with cache reset when the graph generation changes so stale listings never leak across sessions.
> 
> **`NoteEditor`** passes **`resolveFileLink`**, **`resolveFileInfo`**, and **`onFileClick`** into **`MeowdownEditor`**; pill clicks reuse the existing link path (asset opener → deep links → OS URL). **`NotePane`** connects the persistence helpers to the editor.
> 
> Tests cover link vetting, size resolution (single listing, save-seeded sizes, unsafe/missing/session edge cases), and editor prop/click plumbing.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 2dddcd79b127594bebaf36e107c805c38d133f34. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

## Summary by CodeRabbit

* **New Features**
  * Note attachments now render linked non-image assets as compact file pills, including file-size details when available.
  * File-pill clicks route through the same in-app link behavior as other note links (assets open the asset view).
* **Bug Fixes**
  * Attachment link resolution now accepts only safe, graph-relative `assets/...` hrefs and rejects unsafe/remote/traversal patterns.
  * Asset file metadata is cached and deduplicated per session; resolution safely degrades to unknown sizes when sessions end or lookups fail.
* **Tests**
  * Expanded coverage for file-pill prop wiring, opener routing, and asset listing/link resolution edge cases.

### CODE REVIEW

File: apps/desktop/src/editor/use-asset-persistence.ts [hunk 1]
```diff
@@ -174,8 +212,36 @@ export function useAssetPersistence(
...
+        const entries = await pendingAssetListing.current
+        for (const entry of entries) {
+          cache.set(entry.path, entry.size)
+        }
+      }
+      const size = cache.get(href)
+      return size === undefined ? undefined : { size }
+    },
+    [generation],
+  )
```
[the remark this answers is not in the document]
Comment (maccman) [reply apps/desktop/src/editor/use-asset-persistence.ts:247]: Fixed in 2dddcd79: the size map is swapped per generation (not cleared), and both saveFile and resolveFileInfo capture the map reference before their await — a listing or save resolving after a graph switch writes into the orphaned instance. Regression test added.

File: apps/desktop/src/editor/use-asset-persistence.ts [hunk 2]
```diff
@@ -174,8 +212,36 @@ export function useAssetPersistence(
...
+        pendingAssetListing.current ??= listDir('assets', generation).finally(() => {
+          pendingAssetListing.current = null
+        })
+        const entries = await pendingAssetListing.current
+        for (const entry of entries) {
+          cache.set(entry.path, entry.size)
+        }
+      }
+      const size = cache.get(href)
+      return size === undefined ? undefined : { size }
```
[the remark this answers is not in the document]
Comment (maccman) [reply apps/desktop/src/editor/use-asset-persistence.ts:244]: Declining this one: resolveFileInfo runs once per rendered pill (meowdown constructs each pill view once), and concurrent lookups share the in-flight listing. Negative caching would make a file that appears later never get its size within a session — not worth the staleness for a rare re-list.

File: apps/desktop/src/editor/use-asset-persistence.ts [hunk 3]
```diff
@@ -99,6 +128,14 @@ export function useAssetPersistence(
...
   }, [path, generation])
 
+  useEffect(() => {
+    const cache = sizeByAssetPath.current
+    return () => {
+      cache.clear()
+      pendingAssetListing.current = null
+    }
+  }, [generation])
+
```
[the remark this answers is not in the document]
Comment (maccman) [reply apps/desktop/src/editor/use-asset-persistence.ts:140]: Fixed in 2dddcd79 with the map-instance swap you proposed; writes now land in the session-scoped instance captured before the await. Regression test covers a listing resolving after a generation switch.

File: apps/desktop/src/editor/use-asset-persistence.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (maccman) [reply apps/desktop/src/editor/use-asset-persistence.ts:247]: Fixed in 2dddcd79: a failed listing now degrades to undefined (pill without a size), per the documented contract. Note meowdown itself does catch and log resolveFileInfo rejections, so this was never an unhandled rejection — but the contract point stands.
